server/src/main/java/org/elasticsearch/index/engine/NoopEngine.java

```diff
@@ -0,0 +1,326 @@
+ * the Apache License, Version 2.0 (the "License"); you may
+ * not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.elasticsearch.index.engine;
+
+import org.apache.lucene.index.DirectoryReader;
+import org.apache.lucene.index.IndexCommit;
+import org.apache.lucene.index.SegmentInfos;
+import org.elasticsearch.common.Nullable;
+import org.elasticsearch.core.internal.io.IOUtils;
+import org.elasticsearch.index.seqno.LocalCheckpointTracker;
+import org.elasticsearch.index.seqno.SequenceNumbers;
+import org.elasticsearch.index.translog.Translog;
+import org.elasticsearch.index.translog.TranslogConfig;
+import org.elasticsearch.index.translog.TranslogCorruptedException;
+import org.elasticsearch.index.translog.TranslogDeletionPolicy;
+
+import java.io.IOException;
+import java.util.Arrays;
+import java.util.List;
+import java.util.Map;
+import java.util.concurrent.CountDownLatch;
+import java.util.function.BiFunction;
+import java.util.function.LongSupplier;
+import java.util.stream.Stream;
+
+/**
+ * NoopEngine is an engine implementation that does nothing but the bare minimum
+ * required in order to have an engine. All attempts to do something (search,
+ * index, get), throw {@link UnsupportedOperationException}. This does maintain
+ * a translog with a deletion policy so that when flushing, no translog is
+ * retained on disk (setting a retention size and age of 0).
+ *
+ * It's also important to notice that this does list the commits of the Store's
+ * Directory so that the last commit's user data can be read for the historyUUID
+ * and last committed segment info.
+ */
+final class NoopEngine extends Engine {
+
+    private final Translog translog;
+    private final IndexCommit lastCommit;
+    private final LocalCheckpointTracker localCheckpointTracker;
+    private final String historyUUID;
+    private final SegmentInfos lastCommittedSegmentInfos;
+
+    NoopEngine(EngineConfig engineConfig) {
+        super(engineConfig);
+
+        store.incRef();
+        boolean success = false;
+        Translog translog = null;
+
+        try {
+            // The deletion policy for the translog should not keep any translogs around, so the min age/size is set to -1
+            final TranslogDeletionPolicy translogDeletionPolicy = new TranslogDeletionPolicy(-1, -1);
+
+            lastCommittedSegmentInfos = store.readLastCommittedSegmentsInfo();
+            translog = openTranslog(engineConfig, translogDeletionPolicy, engineConfig.getGlobalCheckpointSupplier());
```

Comment:
Do we really need a translog? all we want is to validate that the translog has a the right uuid and that it's empty?

Reply:
We use the translog all over the place in other methods in `Engine`. We can try to encapsulate all of these into different methods (similar to #31213), but I'm not sure what benefit we'd actually get from that. In the next iteration we'll need the translog because we'll need to sync and flush it on engine opening so that there are no operations to be replayed during peer recovery (which I think will still need a translog to retrieve stats about # of ops).

Reply:
> I'm not sure what benefit we'd actually get from that

I looked at these methods and I agree it's on the edge - there are quite a few. I would still prefer we replace them with unsupported operations or noop returns (ex empty sanpshots). This will lock things down and prevent things that aren't supposed to happen - I think that's good. An alternative is to implement a NoopTranslog but that's another rabbit hole.

> we'll need to sync and flush it on engine opening

Why is that? I think it's good to only close indices that have no ongoing indexing (like our plan for frozen index). Regardless - why can't we do the flush / trim when we close the open engine and convert it to a noop engine? 

I can see one thing down the road because we may close an index on recovery where it has broken settings (TBD). In that case I would still prefer to make utilities methods like `Store#associateIndexWithNewTranslog` that work on the folder. Note that this problem isn't solved even if we keep the translog open as we can't index operations from it into the lucene index with the `NoopEngine` nor ship to a recovering shard with `NoopEngine` in it. We assume those don't exist. I think we should discuss this separately.

Reply:
> I think it's good to only close indices that have no ongoing indexing (like our plan for frozen index).

Yes, but even with no ongoing indexing, a translog still remains (due to retention policy)

> Regardless - why can't we do the flush / trim when we close the open engine and convert it to a noop engine?

That would require setting a new retention policy on an existing engine (making a part of InternalEngine mutable which makes me `:(`). In the future though, we could do the retention policy, sync, and flush/trim when the NoopEngine is opened, and then immediately close the translog.

In order to do this though, we'll have to remove the `getTranslog` method from Engine, is that something you want me to do as a precursor to this?